src/Evolution/Systems/GrMhd/ValenciaDivClean/Subcell/TciOnDgGrid.cpp

```diff
@@ -0,0 +1,166 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "Evolution/Systems/GrMhd/ValenciaDivClean/Subcell/TciOnDgGrid.hpp"
+
+#include <cstddef>
+
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tags/TempTensor.hpp"
+#include "DataStructures/Tensor/EagerMath/DotProduct.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "DataStructures/Variables.hpp"
+#include "Evolution/DgSubcell/PerssonTci.hpp"
+#include "Evolution/Systems/GrMhd/ValenciaDivClean/KastaunEtAl.hpp"
+#include "Evolution/Systems/GrMhd/ValenciaDivClean/NewmanHamlin.hpp"
+#include "Evolution/Systems/GrMhd/ValenciaDivClean/PalenzuelaEtAl.hpp"
+#include "Evolution/Systems/GrMhd/ValenciaDivClean/PrimitiveFromConservative.hpp"
+#include "NumericalAlgorithms/Spectral/Mesh.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+
+namespace grmhd::ValenciaDivClean::subcell {
+template <typename RecoveryScheme>
+template <size_t ThermodynamicDim>
+bool TciOnDgGrid<RecoveryScheme>::apply(
+    const gsl::not_null<Variables<hydro::grmhd_tags<DataVector>>*> dg_prim_vars,
+    const Scalar<DataVector>& tilde_d, const Scalar<DataVector>& tilde_tau,
+    const tnsr::i<DataVector, 3, Frame::Inertial>& tilde_s,
+    const tnsr::I<DataVector, 3, Frame::Inertial>& tilde_b,
+    const Scalar<DataVector>& tilde_phi,
+    const tnsr::ii<DataVector, 3, Frame::Inertial>& spatial_metric,
+    const tnsr::II<DataVector, 3, Frame::Inertial>& inv_spatial_metric,
+    const Scalar<DataVector>& sqrt_det_spatial_metric,
+    const EquationsOfState::EquationOfState<true, ThermodynamicDim>& eos,
+    const Mesh<3>& dg_mesh, const TciOptions& tci_options,
+    const double persson_exponent) noexcept {
+  constexpr double persson_tci_epsilon = 1.0e-18;
+  const size_t number_of_points = dg_mesh.number_of_grid_points();
+  Variables<hydro::grmhd_tags<DataVector>> temp_prims(number_of_points);
+
+  // require: tilde_d/sqrt{gamma} >= 0.0 (or some positive user-specified value)
+  if (min(get(tilde_d) / get(sqrt_det_spatial_metric)) <
```

Comment:
Because in practice waiting until the density is actually negative may be too late, depending on the subcell solver.

Yes, but flagging atmosphere as troubled isn't obviously undesirable. I agree that generally it would be nice not to, but I don't see why we should forbid it.